modules/videoio/src/cap_ffmpeg_impl.hpp

```diff
@@ -1410,12 +1434,25 @@ bool CvCapture_FFMPEG::retrieveFrame(int, unsigned char** data, int* step, int*
     if (rawMode)
     {
         AVPacket& p = bsfc ? packet_filtered : packet;
-        *data = p.data;
-        *step = p.size;
-        *width = p.size;
+        if (includeExtraData) {
+            includeExtraData = false;
+            extraDataLen = ic->streams[video_stream]->codec->extradata_size + paddExtraData;
```

Comment:
Please take a look on FFmpeg package filters.
Especially on `h264_mp4toannexb` which is already applied.
You are doing the similar things, but it is unclear "why?"

Reply:
Hi alalek,

Sorry if my explanation above is confusing I was trying to explain two things at once (file writing and extra data required for decoding) and probably failed on both counts.  Forgetting the file writing for a second, there are two reasons for this pull request:
1) Currently `cudacodec::VideoReader() `is unable to decode mp4v codec (mpeg4 contained inside mp4 files one of the formats opencv appears writes to on windows nativly from VideoWriter and the format of the test file highgui/video/big_buck_bunny.mp4 ) because the extra data contained in the header of mp4 file which is required for decoding is not present in the raw bitstream.  I think this is might have been why files with this fourcc code are not currently included in cudacodec (to decode mp4v case CV_FOURCC_MACRO('m', 'p', '4', 'v'): // fallthru or similar would need to be present [here](https://github.com/opencv/opencv_contrib/blob/a26f71313009c93d105151094436eecd4a0990ed/modules/cudacodec/src/ffmpeg_video_source.cpp#L78 )).
2) Using h264 as and example from what I have read and from experiance with several ip camera's the sequence and picture parameter sets (SPS and PPS) which are required for decoding h264 files can be sent either;
a) in band (RTP) just at the begining of a stream, and/or
b) in band (RTP) throughout the stream, and/or
c) in the Session Description returned from RTSP DESCRIBE request e.g. sprop-parameter-sets=Z0LgC5ZUCg/I,aM4BrFSAa
If the parameter sets are only transmitted in the Session Description (c) then cudacodec is unable to decode the stream recieved bitstream.

To fix 1) and 2) I thought that the extra data should be sent at the begining of the first raw frame.  From my testing this does not break any existing code relying on `VideoCapture()` when in "raw" mode.  I also thought that it would be useful to signal the length of this data somewhere so that anyone using the api could trim and/or store it if required (for example when writing a stream over multiple file this data would need to be included at the begining of each file if it is not transmitted in band)

Reply:
It is better to separate PRs with changing of extra data for decoding.

BTW, FFmpeg does not have extra bsf for MPEG4 cuvid decoding:
https://github.com/FFmpeg/FFmpeg/blame/945b2dcc631a891fcc8c911891fd6730a40b8248/libavcodec/cuviddec.c#L1138-L1160

Reply:
Hi Alalek,

> It is better to separate PRs with changing of extra data for decoding.

The extra data required for the RTSP stream's is the same as that required for MPEG4.  If I do a seperate PR for RTSP and MPEG4 then that would just involve removing
```
const bool bMp4MPEG4 = eVideoCodec == AV_CODEC_ID_MPEG4 && ( !strcmp(ic->iformat->long_name, "QuickTime / MOV")
    || !strcmp(ic->iformat->long_name, "FLV (Flash Video)") || !strcmp(ic->iformat->long_name, "Matroska / WebM"));
```

from this PR and placing it into another, which I am happy to do but I do not know if that is what you mean?

> 
> BTW, FFmpeg does not have extra bsf for MPEG4 cuvid decoding:
> https://github.com/FFmpeg/FFmpeg/blame/945b2dcc631a891fcc8c911891fd6730a40b8248/libavcodec/cuviddec.c#L1138-L1160

From the Nvidia Video Codec SDK example for MPEG4 it does not need an additional bsf it just needs the extra info contained in 
`fmtc->streams[iVideoStream]->codecpar->extradata`

> NVIDIA doesn't declare any support for MPEG4: https://developer.nvidia.com/video-encode-and-decode-gpu-support-matrix-new

I am not sure why Nvidia doesn't list it on its support matrix, but it is supported and included in their latest programming guide dated [October 26 2021](https://docs.nvidia.com/video-technologies/video-codec-sdk/nvdec-video-decoder-api-prog-guide/#supported-codecs).

Reply:
> it just needs the extra info contained in fmtc->streams[iVideoStream]->codecpar->extradata

So we should not modify "packets".
At least FFmpeg doesn't do that too:

```
ffmpeg -dump_attachment "extra.data" -i testdata/highgui/video/big_buck_bunny.mp4 -vcodec copy -an -f rawvideo test.mp4v
```

So codec's extradata is a dedicated output for FFmpeg.

---

I believe we should interpret that as dedicated output too and implement this through call:

```
cap.retrieve(codec_extradata_buffer, <video_codec_extradata_idx_property_value>);
```

Reply:
> So we should not modify "packets".

Ideally no, I added to the head of the existing packet to avoid changing the api, and didn't think of using retrieve, which seems like a much better option.

So will it be ok to modify everything down from 
`virtual bool retrieveFrame(int, cv::OutputArray frame) CV_OVERRIDE`
to include `<video_codec_extradata_idx_property_value>`?

Reply:
Just use "int" parameter to retrieve non-frame data.
"0" value corresponds to video stream (default).

Reply:
@alalek updated to address your concerns.